File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def constructBST(nums):
            if not nums:
                return None
            n = len(nums)
            median = nums[n//2]
            
            root = TreeNode(median)
            left = constructBST(nums[:n//2])
            right = constructBST(nums[n//2+1:])
            root.left = left
            root.right = right
            return root
        return constructBST(nums)
    
        

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        int mid = nums.length / 2;
        if(nums.length == 0)return null;
        TreeNode leftSubTree = sortedArrayToBST(Arrays.copyOfRange(nums, 0, mid));
        TreeNode rightSubTree = sortedArrayToBST(Arrays.copyOfRange(nums, mid + 1, nums.length));
        TreeNode root = new TreeNode(nums[mid]);
        root.left = leftSubTree;
        root.right = rightSubTree;
        return root;
    }
}